README.md
# badi_date

A library to convert Dart DateTime to Badi date and back.


## Usage Example

```dart
import 'package:badi_date/badi_date.dart';

final date1 = BadiDate(day: 1, month: 1, year: 178);
final date2 = BadiDate.fromDate(DateTime.now());

final DateTime start = date1.startDateTime;
final DateTime end = date1.endDateTime;
final BahaiHolyDayEnum holyDay = date1.holyDay;
final bool = date2.isPeriodOfFast;
final bool = date2.isAyyamIHa;
final BadiDate nextHolyDay = date2.nextHolyDate;

```


CHANGELOG.md
## 1.0.4

* Fix issue with Vahid been off by one when it's the last year of the Vahid.
## 1.0.3

* Fix issue when coordinates are set leading to off by one day dates.
## 1.0.2

* Update dart and dependency version.

## 1.0.1

* Update dart and dependency version.

## 1.0.0

* Initial Open Source release.


test/single_test.dart
import 'package:badi_date/badi_date.dart';
import 'package:test/test.dart';

void main() {
  test('it', () {
    var date = DateTime(2021, 1, 17, 18);
    var badiDate = BadiDate.fromDate(date, longitude: 10.0, latitude: 53.6);
    expect(badiDate.startDateTime, DateTime(2021, 1, 17, 16, 34));
    expect(badiDate.endDateTime, DateTime(2021, 1, 18, 16, 36));
  });
}


test/badi_date_test.dart
import 'package:badi_date/badi_date.dart';
import 'package:badi_date/bahai_holyday.dart';
import 'package:test/test.dart';

void main() {
  group('throws errors', () {
    test('Throws for invalid day and month', () {
      expect(() => BadiDate(year: 1, month: -1, day: 1), throwsArgumentError);
      expect(() => BadiDate(year: 1, month: 20, day: 1), throwsArgumentError);
      expect(() => BadiDate(year: 1, month: 1, day: 0), throwsArgumentError);
      expect(() => BadiDate(year: 1, month: 1, day: 20), throwsArgumentError);
      expect(() => BadiDate(year: 1, month: 1, day: 4, ayyamIHa: true),
          throwsArgumentError);
    });

    test('Throws unsupported years', () {
      expect(
          () => BadiDate(year: 222, month: 1, day: 1), throwsUnsupportedError);
      expect(() => BadiDate.fromDate(DateTime(2065, 3, 22)),
          throwsUnsupportedError);
    });
  });

  group('date today', () {
    final badiDate = BadiDate(day: 6, month: 16, year: 177);
    test('year, month, day', () {
      final expected = DateTime(2021, 1, 4);
      expect(badiDate.startDateTime.year, equals(expected.year));
      expect(badiDate.startDateTime.month, equals(expected.month));
      expect(badiDate.endDateTime.day, equals(expected.day));
    });

    test('after sunset', () {
      final expected = DateTime(2021, 1, 3);
      expect(badiDate.startDateTime.year, equals(expected.year));
      expect(badiDate.startDateTime.month, equals(expected.month));
      expect(badiDate.startDateTime.day, equals(expected.day));
    });

    test('year in Vahid, Vahid, Kull-i-shay', () {
      expect(badiDate.kullIShay, equals(1));
      expect(badiDate.vahid, equals(10));
      expect(badiDate.yearInVahid, equals(6));
    });

    test('feast, fast, Ayyam-i-ha, Holy day', () {
      expect(badiDate.isAyyamIHa, equals(false));
      expect(badiDate.isFeastDay, equals(false));
      expect(badiDate.isPeriodOfFast, equals(false));
      expect(badiDate.holyDay, equals(null));
    });

    test('next feast, next holy day, day of the year', () {
      expect(badiDate.nextHolyDate.hashCode, equals(178001));
      expect(BadiDate(year: 178, month: 1, day: 1).hashCode, equals(178001));

      expect(
          badiDate.nextHolyDate, equals(BadiDate(year: 178, month: 1, day: 1)));
      expect(badiDate.getNextFeast(),
          equals(BadiDate(year: 177, month: 17, day: 1)));
      expect(badiDate.dayOfYear, equals(291));
    });

    test('last AyyamIHa day', () {
      expect(badiDate.lastAyyamIHaDayOfYear.endDateTime,
          equals(DateTime(2021, 2, 28, 18)));
      expect(badiDate.lastAyyamIHaDayOfYear.startDateTime,
          equals(DateTime(2021, 2, 27, 18)));
    });
  });

  group('naw ruz', () {
    test('naw ruz in gregorian date', () {
      for (int i = 1; i < 175; i++) {
        final date = BadiDate(day: 1, month: 1, year: i);
        expect(date.holyDay, equals(BahaiHolyDayEnum.NAW_RUZ));
        expect(date.startDateTime.month, equals(3));
        expect(date.startDateTime.year, equals(i + 1843));
        if (i < 173) {
          expect(date.endDateTime.day, equals(21), reason: 'Year $i');
          expect(BadiDate.getDayOfNawRuz(i), equals(21), reason: 'Year $i');
        } else {
          expect(date.endDateTime.day, equals(20), reason: 'Year $i');
          expect(BadiDate.getDayOfNawRuz(i), equals(20), reason: 'Year $i');
        }
      }
    });
  });

  group('feasts and holy days', () {
    final badiDate = BadiDate(day: 1, month: 19, year: 177);
    test('calculates the feasts', () {
      expect(badiDate.endDateTime, equals(DateTime(2021, 3, 1, 18)));
      BadiDate date = badiDate.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 3, 20, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 4, 8, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 4, 27, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 5, 16, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 6, 4, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 6, 23, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 7, 12, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 7, 31, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 8, 19, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 9, 7, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 9, 26, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 10, 15, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 11, 3, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 11, 22, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 12, 11, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2021, 12, 30, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2022, 1, 18, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2022, 2, 6, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2022, 3, 2, 18)));
      date = date.getNextFeast();
      expect(date.endDateTime, equals(DateTime(2022, 3, 21, 18)));
    });

    test('calculates the holy days', () {
      BadiDate date = badiDate.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 3, 20, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.NAW_RUZ));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 4, 20, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.RIDVAN1ST));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 4, 28, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.RIDVAN9TH));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 5, 1, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.RIDVAN12TH));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 5, 23, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.DECLARATION_OF_THE_BAB));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 5, 28, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.ASCENSION_OF_BAHAULLAH));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 7, 9, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.MARTYRDOM_OF_THE_BAB));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 11, 6, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.BIRTH_OF_THE_BAB));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 11, 7, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 11, 25, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.DAY_OF_THE_COVENANT));

      date = date.nextHolyDate;
      expect(date.endDateTime, equals(DateTime(2021, 11, 27, 18)));
      expect(date.holyDay, equals(BahaiHolyDayEnum.ASCENSION_OF_ABDUL_BAHA));
    });
  });

  group('corner cases', () {
    test('round trip naw ruz', () {
      var date = DateTime(2021, 3, 19, 20);
      var badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2021, 3, 19, 18));
      expect(badiDate.endDateTime, DateTime(2021, 3, 20, 18));
      expect(badiDate.month, 1);
      expect(badiDate.day, 1);
      expect(badiDate.year, 178);

      date = DateTime(2021, 3, 19, 16);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2021, 3, 18, 18));
      expect(badiDate.endDateTime, DateTime(2021, 3, 19, 18));
      expect(badiDate.month, 19);
      expect(badiDate.day, 19);
      expect(badiDate.year, 177);

      date = DateTime(2021, 3, 20, 16);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2021, 3, 19, 18));
      expect(badiDate.endDateTime, DateTime(2021, 3, 20, 18));
      expect(badiDate.month, 1);
      expect(badiDate.day, 1);
      expect(badiDate.year, 178);
    });

    test('new years eve', () {
      var date = DateTime(2020, 12, 31, 16);
      var badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2020, 12, 30, 18));
      expect(badiDate.endDateTime, DateTime(2020, 12, 31, 18));
      expect(badiDate.day, 2);
      expect(badiDate.month, 16);
      expect(badiDate.year, 177);

      date = DateTime(2020, 12, 31, 20);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2020, 12, 31, 18));
      expect(badiDate.endDateTime, DateTime(2021, 1, 1, 18));
      expect(badiDate.day, 3);
      expect(badiDate.month, 16);
      expect(badiDate.year, 177);

      date = DateTime(2021, 1, 1, 16);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.startDateTime, DateTime(2020, 12, 31, 18));
      expect(badiDate.endDateTime, DateTime(2021, 1, 1, 18));
      expect(badiDate.day, 3);
      expect(badiDate.month, 16);
      expect(badiDate.year, 177);
    });

    test('leap day', () {
      var date = DateTime(2020, 2, 29);
      var badiDate = BadiDate.fromDate(date);
      expect(badiDate.year, 176);
      expect(badiDate.isAyyamIHa, true);
      expect(badiDate.isPeriodOfFast, false);
      expect(badiDate.day, 4);

      date = DateTime(2020, 3, 1);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.isAyyamIHa, false);
      expect(badiDate.isPeriodOfFast, true);
      expect(badiDate.day, 1);

      date = DateTime(2021, 2, 28);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.isAyyamIHa, true);
      expect(badiDate.isPeriodOfFast, false);
      expect(badiDate.day, 4);

      date = DateTime(2021, 3, 1);
      badiDate = BadiDate.fromDate(date);
      expect(badiDate.isAyyamIHa, false);
      expect(badiDate.isPeriodOfFast, true);
      expect(badiDate.day, 1);
    });

    for (final tc in <_VahidTestCase>[
      _VahidTestCase(
        year: 2013,
        expectedVahid: 9,
        expectedYearInVahid: 18,
      ),
      _VahidTestCase(
        year: 2014,
        expectedVahid: 9,
        expectedYearInVahid: 19,
      ),
      _VahidTestCase(
        year: 2015,
        expectedVahid: 10,
        expectedYearInVahid: 1,
      ),
    ]) {
      test('vahid for year ${tc.year}', () {
        var date = DateTime(tc.year, 10, 21, 12);
        var badiDate = BadiDate.fromDate(date);
        expect(badiDate.vahid, tc.expectedVahid);
        expect(badiDate.yearInVahid, tc.expectedYearInVahid);
      });
    }

    test('kull i shay', () {
      var date = DateTime(2023, 10, 21, 12);
      var badiDate = BadiDate.fromDate(date);
      expect(badiDate.kullIShay, 1);
      expect(badiDate.vahid, 10);
      expect(badiDate.yearInVahid, 9);
    });
  });

  group('sunset calculation', () {
    test("handles pols and nonsense values", () {
      final expected = DateTime(2021, 1, 17, 18);
      expect(BadiDate(day: 1, month: 17, year: 177).startDateTime,
          equals(expected));
      expect(
          BadiDate(day: 1, month: 17, year: 177, latitude: 53.6).startDateTime,
          equals(expected));
      expect(
          BadiDate(day: 1, month: 17, year: 177, longitude: 10.0).startDateTime,
          equals(expected));
      expect(
          BadiDate(
                  day: 1,
                  month: 17,
                  year: 177,
                  longitude: 190.0,
                  latitude: 53.6)
              .startDateTime,
          equals(expected));
      expect(
          BadiDate(
                  day: 1,
                  month: 17,
                  year: 177,
                  longitude: -190.0,
                  latitude: 53.6)
              .startDateTime,
          equals(expected));
      expect(
          BadiDate(
                  day: 1, month: 17, year: 177, longitude: 10.0, latitude: 66.6)
              .startDateTime,
          equals(expected));
      expect(
          BadiDate(
                  day: 1,
                  month: 17,
                  year: 177,
                  longitude: 10.0,
                  latitude: -66.6)
              .startDateTime,
          equals(expected));
    });

    // This test only works if your local time is CET or CEST
    test("calculates sunset", () {
      final expected = DateTime.utc(2021, 1, 17, 15, 34);
      expect(
          BadiDate(
                  day: 1, month: 17, year: 177, longitude: 10.0, latitude: 53.6)
              .startDateTime
              .toUtc(),
          equals(expected));
    });

    // This test only works if your local time is CET or CEST
    test("handles daylight saving", () {
      final expected = DateTime.utc(2021, 6, 23, 19, 55);
      expect(
          BadiDate(day: 1, month: 6, year: 178, longitude: 10.0, latitude: 53.6)
              .endDateTime
              .toUtc(),
          equals(expected));
    });

    test("handles sunset on next Gregorian day", () {
      final expected = DateTime.utc(2021, 6, 23, 22, 18);
      expect(
          BadiDate(day: 1, month: 6, year: 178, longitude: 8.0, latitude: 64.6)
              .endDateTime
              .toUtc(),
          equals(expected));
    });

    test("handles dates in far west", () {
      final badiDate =
          BadiDate(day: 2, month: 11, year: 179, longitude: -86, latitude: 12);
      BadiDate date = badiDate.nextHolyDate;
      expect(date.holyDay, equals(BahaiHolyDayEnum.BIRTH_OF_THE_BAB));
      expect(
          date.endDateTime.toUtc(), equals(DateTime.utc(2022, 10, 26, 23, 21)));
      expect(date.startDateTime.toUtc(),
          equals(DateTime.utc(2022, 10, 25, 23, 22)));
    });

    test("handles dates in far east", () {
      final badiDate =
          BadiDate(day: 2, month: 11, year: 179, longitude: 179, latitude: -12);
      BadiDate date = badiDate.nextHolyDate;
      expect(date.holyDay, equals(BahaiHolyDayEnum.BIRTH_OF_THE_BAB));
      expect(
          date.endDateTime.toUtc(), equals(DateTime.utc(2022, 10, 26, 6, 3)));
      expect(
          date.startDateTime.toUtc(), equals(DateTime.utc(2022, 10, 25, 6, 3)));
    });
  });
}

class _VahidTestCase {
  _VahidTestCase({
    required this.year,
    required this.expectedVahid,
    required this.expectedYearInVahid,
  });
  final int year;
  final int expectedVahid;
  final int expectedYearInVahid;
}


lib/names.dart
const monthNames = {
  1: "Bahá",
  2: "Jalál",
  3: "Jamál",
  4: "‘Aẓamat",
  5: "Núr",
  6: "Raḥmat",
  7: "Kalimát",
  8: "Kamál",
  9: "Asmá’",
  10: "‘Izzat",
  11: "Mashíyyat",
  12: "‘Ilm",
  13: "Qudrat",
  14: "Qawl",
  15: "Masá’il",
  16: "Sharaf",
  17: "Sulṭán",
  18: "Mulk",
  19: "‘Alá’"
};

const vahid = "Váḥid";
const kullIShay = "Kull-i-Shay’";
const ayyamIHa = "Ayyám-i-Há";

const weekDay = {
  1: "Jamál",
  2: "Kamál",
  3: "Fiḍál",
  4: "‘Idál",
  5: "Istijlál",
  6: "Istiqlál",
  7: "Jalál"
};

const vahids = {
  1: "Alif",
  2: "Bá’",
  3: "Ab",
  4: "Dál",
  5: "Báb",
  6: "Váv",
  7: "Abad",
  8: "Jád",
  9: "Bahá’",
  10: "Ḥubb",
  11: "Bahháj",
  12: "Javáb",
  13: "Aḥad",
  14: "Vahháb",
  15: "Vidád",
  16: "Badí‘",
  17: "Bahí",
  18: "Abhá",
  19: "Váḥid"
};


lib/years.dart
class YearSpecifics {
  final int year;
  final bool leapday;
  final bool nawRuzOnMarch21;
  final int birthOfBab;

  const YearSpecifics(
      {required this.year,
      required this.birthOfBab,
      this.leapday = false,
      this.nawRuzOnMarch21 = false});
}

const yearSpecifics = {
  172: YearSpecifics(year: 172, birthOfBab: 238, nawRuzOnMarch21: true),
  173: YearSpecifics(year: 173, birthOfBab: 227),
  174: YearSpecifics(year: 174, birthOfBab: 216, leapday: true),
  175: YearSpecifics(year: 175, birthOfBab: 234, nawRuzOnMarch21: true),
  176: YearSpecifics(year: 176, birthOfBab: 223, nawRuzOnMarch21: true),
  177: YearSpecifics(year: 177, birthOfBab: 213),
  178: YearSpecifics(year: 178, birthOfBab: 232, leapday: true),
  179: YearSpecifics(year: 179, birthOfBab: 220, nawRuzOnMarch21: true),
  180: YearSpecifics(year: 180, birthOfBab: 210, nawRuzOnMarch21: true),
  181: YearSpecifics(year: 181, birthOfBab: 228),
  182: YearSpecifics(year: 182, birthOfBab: 217, leapday: true),
  183: YearSpecifics(year: 183, birthOfBab: 235, nawRuzOnMarch21: true),
  184: YearSpecifics(year: 184, birthOfBab: 224, nawRuzOnMarch21: true),
  185: YearSpecifics(year: 185, birthOfBab: 214),
  186: YearSpecifics(year: 186, birthOfBab: 233),
  187: YearSpecifics(year: 187, birthOfBab: 223, leapday: true),
  188: YearSpecifics(year: 188, birthOfBab: 211, nawRuzOnMarch21: true),
  189: YearSpecifics(year: 189, birthOfBab: 230),
  190: YearSpecifics(year: 190, birthOfBab: 238),
  191: YearSpecifics(year: 191, birthOfBab: 238, leapday: true),
  192: YearSpecifics(year: 192, birthOfBab: 226, nawRuzOnMarch21: true),
  193: YearSpecifics(year: 193, birthOfBab: 215),
  194: YearSpecifics(year: 194, birthOfBab: 234),
  195: YearSpecifics(year: 195, birthOfBab: 224, leapday: true),
  196: YearSpecifics(year: 196, birthOfBab: 213, nawRuzOnMarch21: true),
  197: YearSpecifics(year: 197, birthOfBab: 232),
  198: YearSpecifics(year: 198, birthOfBab: 221),
  199: YearSpecifics(year: 199, birthOfBab: 210, leapday: true),
  200: YearSpecifics(year: 200, birthOfBab: 228, nawRuzOnMarch21: true),
  201: YearSpecifics(year: 201, birthOfBab: 217),
  202: YearSpecifics(year: 202, birthOfBab: 236),
  203: YearSpecifics(year: 203, birthOfBab: 225, leapday: true),
  204: YearSpecifics(year: 204, birthOfBab: 214, nawRuzOnMarch21: true),
  205: YearSpecifics(year: 205, birthOfBab: 233),
  206: YearSpecifics(year: 206, birthOfBab: 223),
  207: YearSpecifics(year: 207, birthOfBab: 212, leapday: true),
  208: YearSpecifics(year: 208, birthOfBab: 230, nawRuzOnMarch21: true),
  209: YearSpecifics(year: 209, birthOfBab: 219),
  210: YearSpecifics(year: 210, birthOfBab: 237),
  211: YearSpecifics(year: 211, birthOfBab: 227, leapday: true),
  212: YearSpecifics(year: 212, birthOfBab: 215, nawRuzOnMarch21: true),
  213: YearSpecifics(year: 213, birthOfBab: 234),
  214: YearSpecifics(year: 214, birthOfBab: 224),
  215: YearSpecifics(year: 215, birthOfBab: 213),
  216: YearSpecifics(year: 216, birthOfBab: 232, leapday: true),
  217: YearSpecifics(year: 217, birthOfBab: 220),
  218: YearSpecifics(year: 218, birthOfBab: 209),
  219: YearSpecifics(year: 219, birthOfBab: 228),
  220: YearSpecifics(year: 220, birthOfBab: 218, leapday: true),
  221: YearSpecifics(year: 221, birthOfBab: 236),
};


lib/bahai_holyday.dart
enum BahaiHolyDayEnum {
  /// Naw-Ruz
  NAW_RUZ,

  /// 1st day of Ridvan
  RIDVAN1ST,

  /// 9th day of Ridvan
  RIDVAN9TH,

  /// 12th day of Ridvan
  RIDVAN12TH,

  /// Declaration of the Bab
  DECLARATION_OF_THE_BAB,

  /// Ascension of Baha'u'llah
  ASCENSION_OF_BAHAULLAH,

  /// Martyrdom of the Bab
  MARTYRDOM_OF_THE_BAB,

  /// Birth of the Bab
  BIRTH_OF_THE_BAB,

  /// Birth of Baha'u'llah
  BIRTH_OF_BAHAULLAH,

  /// Day of the Covenant (not a holy day)
  DAY_OF_THE_COVENANT,

  /// Ascension of Abdu'l-Baha (not a holy day)
  ASCENSION_OF_ABDUL_BAHA,
}

class BahaiHolyDay {
  final int _dayOfYear;
  final BahaiHolyDayEnum type;

  const BahaiHolyDay(this._dayOfYear, this.type);

  int getDayOfTheYear({int? dayOfYearBirthOfBab}) {
    if (type == BahaiHolyDayEnum.BIRTH_OF_THE_BAB) {
      return dayOfYearBirthOfBab ?? _dayOfYear;
    } else if (type == BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH &&
        dayOfYearBirthOfBab != null) {
      return dayOfYearBirthOfBab + 1;
    }
    return _dayOfYear;
  }
}

const List<BahaiHolyDay?> bahaiHolyDays = [
  /// Naw-Ruz
  BahaiHolyDay(1, BahaiHolyDayEnum.NAW_RUZ),

  /// 1st day of Ridvan
  BahaiHolyDay(32, BahaiHolyDayEnum.RIDVAN1ST),

  /// 9th day of Ridvan
  BahaiHolyDay(40, BahaiHolyDayEnum.RIDVAN9TH),

  /// 12th day of Ridvan
  BahaiHolyDay(43, BahaiHolyDayEnum.RIDVAN12TH),

  /// Decleration of the Bab
  BahaiHolyDay(65, BahaiHolyDayEnum.DECLARATION_OF_THE_BAB),

  /// Ascension of Baha'u'llah
  BahaiHolyDay(70, BahaiHolyDayEnum.ASCENSION_OF_BAHAULLAH),

  /// Martyrdom of the Bab
  BahaiHolyDay(112, BahaiHolyDayEnum.MARTYRDOM_OF_THE_BAB),

  /// Birth of the Bab
  BahaiHolyDay(214, BahaiHolyDayEnum.BIRTH_OF_THE_BAB),

  /// Birth of Baha'u'llah
  BahaiHolyDay(237, BahaiHolyDayEnum.BIRTH_OF_BAHAULLAH),

  /// Day of the Covenant (not a holy day)
  BahaiHolyDay(251, BahaiHolyDayEnum.DAY_OF_THE_COVENANT),

  /// Ascension of Abdu'l-Baha (not a holy day)
  BahaiHolyDay(253, BahaiHolyDayEnum.ASCENSION_OF_ABDUL_BAHA),
];
